### ISSUE TITLE
Unable to override CustomConversions for Couchbase

### ISSUE THREAD

**Issue by edudar** [spring-projects/spring-boot#7700]
I see that this feature was supposed to be added in #5979 but the framework code has changed since then and I can't get it to work, no matter what I do.

I have `data-couchbase` starter as a dependency
```
compile 'org.springframework.boot:spring-boot-starter-data-couchbase:1.4.2.RELEASE'
```

I also have a quite simple configuration class
```
@Configuration
@EnableCouchbaseRepositories
public class ServiceContext {
    @Bean(name = BeanNames.COUCHBASE_CUSTOM_CONVERSIONS)
    CustomConversions customConversions() {
        return new CustomConversions(...);
    }
}
```

with main class as simple as 
```
@SpringBootApplication
public class Service {
    public static void main(String[] args) { SpringApplication.run(Service.class, args); }
}
```

I expect that spring-boot takes care of autoconfiguration and it actually does. When I start application this bean never gets created though, execution always go to `AbstractCouchbaseDataConfiguration.customConversions()` through its subclass - `SpringBootCouchbaseDataConfiguration`. When I change the name of this bean just to see what will happen, `AbstractCouchbaseConverter.afterPropertiesSet()` gets called way before code gets to my custom bean drawing it essentially useless. 

Either I miss something very big and obvious or there's a bug/regression going on...

**Comment by snicoll** [comment]
> I see that this feature was supposed to be added in #5979 

No it wasn't. If you look at the comment, you'll see @simonbasle states it's not needed (and no code was added).

FYI `@EnableCouchbaseRepositories` in your configuration is useless, Spring Boot does it for you.

As for the original report, I am afraid this has nothing to do with Spring Boot (a spring data couchbase glitch maybe?) but I'll have a look (btw it would be nicer to share a project that we can actually run rather than copying the code in a comment).

**Comment by edudar** [comment]
Here's demo repo that showcases the issue: https://github.com/edudar/spring-boot-7700. 

What I've found is that removing/commenting-out
```
spring:
  couchbase:
    bootstrap-hosts: localhost
```
from `application.yml` resolves this somehow.

**Comment by edudar** [comment]
Well, guess it just disables this whole spring-boot auto config for couchbase when no bootstrap hosts are available and that's why initialization comes to custom bean. Definitely not the workaround of any sort.

**Comment by snicoll** [comment]
You did. There is nothing in the Spring Boot doc about custom converter and the [`spring-data-couchbase` documentation states that you should override `AbstractCouchbaseConfiguration`](http://docs.spring.io/spring-data/couchbase/docs/current/reference/html/#datatypes)

I don't know what feature you're referring to here but it hasn't been implemented yet. Having said that, Spring Boot could detect such bean in the context.

**Comment by snicoll** [comment]
@edudar having said that, I would raise an issue with `spring-data-couchbase`. Maybe the default implementation should look for converted in the context rather than asking you to extend from a base configuration.

**Comment by edudar** [comment]
@snicoll hmmm, I'm referring to the point 2 in #5979 that says that it should be possible to override `customConversions`. It is a ticket in `spring-boot` and not in `spring-data-couchbase`. Without `spring-boot`, yes, I can extend either `AbstractCouchbaseConfiguration` or even `AbstractCouchbaseDataConfiguration` and it will work but in the case of `spring-boot` there is `SpringBootCouchbaseDataConfiguration` which looks like a boot replacement for `AbstractCouchbaseConfiguration ` (am I wrong?) but also it's package-private so I can't extend from it nicely.

**Comment by edudar** [comment]
Okay, I think I got it. That ticket #5979 most likely is about 1.3.x way of doing Couchbase with Spring Boot and then it perfectly makes sense as I did that in the past. But as we use 1.4.x and Couchbase became a first-class citizen, things changed and `SpringBootCouchbaseDataConfiguration` replaced `AbstractCouchbaseConfiguration`. However, some features, like `customConversions` were not ported to 1.4.x. so it's either 1.3.x and `AbstractCouchbaseConfiguration` or the other workaround I implemented for now.

**Comment by simonbasle** [comment]
chiming in as I made this comment. yes, the context is outdated.

the possibility of declaring a same-name bean in another `Configuration` was never promoted to a fully fledged public feature as far as I remember, and the comment you linked to was referring to something that didn't get merged in 1.4.x in the end.
At some intermediary point it must have worked but without being an official target, and the later changes in couchbase configuration "hierarchy" changed that (while still focusing on the official way of having to redefine `AbstractCouchbaseConfiguration` yourself if you want "advanced" configuration).

The whole `CustomConversions` feels kind of legacy in hindsight though. It can probably be updated with features like automatic registration of `Converter` beans from the context. Don't hesitate to open an issue accordingly on https://jira.spring.io/browse/DATACOUCH/
